File: ios/MoodleLMS/Core/Auth/AuthRepository.swift
import Foundation

final class AuthRepository {
    private let client: MoodleClient
    private let tokenStore: KeychainStore

    static let devToken = "c269d73b8ec3265227714bf37f4dd2e4"
    static let devUserId = 1003

    init(client: MoodleClient, tokenStore: KeychainStore) {
        self.client = client
        self.tokenStore = tokenStore
    }

    func login(username: String, password: String, userId: Int) async throws {
        let token = try await client.login(username: username, password: password)
        tokenStore.save(token: token, userId: userId)
    }

    func useDevToken() {
        tokenStore.save(token: Self.devToken, userId: Self.devUserId)
    }
}

File: android/app/src/main/java/io/itcorner/moodle/core/auth/AuthRepository.kt
package io.itcorner.moodle.core.auth

import io.itcorner.moodle.core.network.MoodleApi
import io.itcorner.moodle.core.network.MoodleApiException
import java.io.IOException
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class AuthRepository @Inject constructor(
    private val api: MoodleApi,
    private val tokenStore: TokenStore,
) {
    fun isLoggedIn(): Boolean = tokenStore.token() != null

    fun useDevToken() {
        tokenStore.save(token = DEV_TOKEN, userId = DEV_USER_ID)
    }

    suspend fun login(username: String, password: String, userId: Long): Result<Unit> {
        return try {
            val response = api.login(username, password)
            val token = response.token
            if (token.isNullOrBlank()) {
                Result.failure(
                    MoodleApiException(
                        errorCode = response.errorcode,
                        message = response.error ?: "Invalid credentials",
                    )
                )
            } else {
                tokenStore.save(token, userId)
                Result.success(Unit)
            }
        } catch (e: IOException) {
            Result.failure(e)
        } catch (e: MoodleApiException) {
            Result.failure(e)
        }
    }

    fun logout() = tokenStore.clear()

    companion object {
        const val DEV_TOKEN = "c269d73b8ec3265227714bf37f4dd2e4"
        const val DEV_USER_ID = 1003L
    }
}
